### PR TITLE
Bd 340 - Zapper contract

### PR DESCRIPTION
**Pull request by RiccardoGalbusera**
- write Zapper.sol contract
- test contract's main functions: zapIn and zapOut

### CODE REVIEW

File: contracts/utils/Zapper.sol [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+
+        (, int24 tickPool, , , , , ) = IUniswapV3Pool(
+            NFTHelper._getPoolAddress(uniswapAddressHolder.uniswapV3FactoryAddress(), token0, token1, fee)
+        ).slot0();
+        uint256 ratioE18 = SwapHelper.getRatioFromRange(tickPool, tickLower, tickUpper);
+        uint256 amountInTo0 = (amountIn * 1e18) / (ratioE18 + 1e18);
+        uint256 amountInTo1 = amountIn - amountInTo0;
+
+        ERC20Helper._approveToken(tokenIn, address(swapRouter), amountIn);
+        if (tokenIn != token0) {
```
Comment (AlbyGiaco) [inline contracts/utils/Zapper.sol:60]: Maybe ternary operator ( x ? y : z ) instead of this if

File: contracts/utils/Zapper.sol [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+
+        ERC20Helper._approveToken(tokenOut, address(this), amount0 + amount1);
+        ERC20Helper._withdrawTokens(tokenOut, msg.sender, amount0 + amount1);
+    }
+
+    ///@notice orders token addresses
+    ///@param token0 address of first token
+    ///@param token1 address of second token
+    ///@return address ordered token address
+    ///@return address ordered token address
```
Comment (luduvigo) [inline contracts/utils/Zapper.sol:182]: Duplicated line

File: contracts/utils/Zapper.sol [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+        uint16[4] memory fees = [100, 500, 3000, 10000];
+
+        for (uint8 i = 0; i < 4; i++) {
+            try this.getPoolLiquidity(token0, token1, uint24(fees[i])) returns (uint128 nextLiquidity) {
+                if (nextLiquidity > bestLiquidity) {
+                    bestLiquidity = nextLiquidity;
+                    fee = fees[i];
+                }
+            } catch {
+                //pass
```
Comment (luduvigo) [inline contracts/utils/Zapper.sol:206]: What happens if we get into this catch block?

File: contracts/utils/Zapper.sol (hunk 2 above)
Comment (RiccardoGalbusera) [reply contracts/utils/Zapper.sol:182]: The function has two outputs, it's correct that there are two lines, but it's true that I should diversify the comments

File: contracts/utils/Zapper.sol (hunk 3 above)
Comment (RiccardoGalbusera) [reply contracts/utils/Zapper.sol:206]: Nothing! If a pool with those tokens and fee doesn't exist, we can't use that and we check for the next one. If no pool exists, than bestLiquidity is never updated and we end in the if block => the transaction reverts

**Review by luduvigo — APPROVED** [review]

**Pull request merged**